host/src/lib.rs

```diff
@@ -441,21 +445,43 @@ impl ScalarUDFImpl for WasmScalarUdf {
         })
     }
 
-    fn invoke_with_args(&self, args: ScalarFunctionArgs) -> DataFusionResult<ColumnarValue> {
-        async_in_sync_context(async {
-            let args = args.try_into()?;
-            let mut store_guard = self.store.lock().await;
-            let return_type = self
-                .bindings
-                .datafusion_udf_wasm_udf_types()
-                .scalar_udf()
-                .call_invoke_with_args(store_guard.deref_mut(), self.resource, &args)
-                .await
-                .context(
-                    "call ScalarUdf::invoke_with_args",
-                    Some(&store_guard.data().stderr.contents()),
-                )??;
-            return_type.try_into()
-        })
+    fn invoke_with_args(&self, _args: ScalarFunctionArgs) -> DataFusionResult<ColumnarValue> {
+        Err(DataFusionError::NotImplemented(
+            "synchronous invocation of WasmScalarUdf is not supported, use invoke_async_with_args instead".to_string(),
+        ))
+    }
+}
+
+#[async_trait]
+impl AsyncScalarUDFImpl for WasmScalarUdf {
+    fn ideal_batch_size(&self) -> Option<usize> {
+        None
+    }
+
+    async fn invoke_async_with_args(
+        &self,
+        args: ScalarFunctionArgs,
+        _option: &ConfigOptions,
+    ) -> DataFusionResult<arrow::array::ArrayRef> {
+        let args = args.try_into()?;
+        let mut store_guard = self.store.lock().await;
+        let return_type = self
+            .bindings
+            .datafusion_udf_wasm_udf_types()
+            .scalar_udf()
+            .call_invoke_with_args(store_guard.deref_mut(), self.resource, &args)
+            .await
+            .context(
+                "call ScalarUdf::invoke_with_args",
+                Some(&store_guard.data().stderr.contents()),
+            )??;
+
+        drop(store_guard);
+
+        let columnar_value: ColumnarValue = return_type.try_into()?;
+        match columnar_value {
+            ColumnarValue::Array(v) => Ok(v),
+            ColumnarValue::Scalar(v) => v.to_array_of_size(1),
```

Comment:
The output row count should be the same as the input. The `args` (IIRC) should have the respective row in them, so you can use that.